source/SH3/system/glcontext.cpp

```diff
@@ -17,16 +17,122 @@
+ *
+ *  Used in conjunction with glDebugMessageCallback
+ *
+ *  @arg source     The Source of this error
+ *  @arg type       The type of error this is
+ *  @arg id         Error ID
+ *  @arg severity   How severe this error is. If it's fatal, terminate the program with @ref die()
+ *  @arg length     Length of the message log
+ *  @arg message    Error message passed to us from OpenGL
+ *  @arg data       User Param data
+ */
+static void GLAPIENTRY debugCallback(GLenum source, GLenum type, GLuint id, GLenum severity, GLsizei length, const GLchar* message, const GLvoid* data)
+{   
+    LogLevel    level;
+    std::string errMsg;
+    
+    static_cast<void>(id);
+    static_cast<void>(length);
+    
+    ASSERT(!data);
+    errMsg = "";
+
+    // Get the source of the error message
+    switch(source)
+    {
+    case GL_DEBUG_SOURCE_API: // Error caused by the OpenGL API
+        errMsg += "API ";
+        break;
+    case GL_DEBUG_SOURCE_WINDOW_SYSTEM:
+        errMsg += "WINDOW SYSTEM ";
+        break;
+    case GL_DEBUG_SOURCE_SHADER_COMPILER:
+        errMsg += "SHADER COMPILER ";
+        break;
+    case GL_DEBUG_SOURCE_APPLICATION:
+        errMsg += "APPLICATION ";
+        break;
+    case GL_DEBUG_SOURCE_OTHER:
+        errMsg += "OTHER ";
+        break;
+    default:
+        errMsg += "UNKNOWN ";
+        break;
+    }
+
+    // Get the type of error this is
+    switch(type)
+    {
+    case GL_DEBUG_TYPE_ERROR:
+        errMsg += "(ERROR): ";
+        break;
+    case GL_DEBUG_TYPE_DEPRECATED_BEHAVIOR:
+        errMsg += "(DEPRECATED): ";
+        break;
+    case GL_DEBUG_TYPE_UNDEFINED_BEHAVIOR:
+        errMsg += "(UNDEFINED BEHAVIOR): ";
+        break;
+    case GL_DEBUG_TYPE_PORTABILITY:
+        errMsg += "PORTABILITY! ";
+        break;
+    case GL_DEBUG_TYPE_PERFORMANCE:
+        errMsg += "(PERFORMANCE): ";
+        break;
+    case GL_DEBUG_TYPE_OTHER:
+        errMsg += "(OTHER): ";
+        break;
+    case GL_DEBUG_TYPE_MARKER:
+        errMsg += "(MARKER): ";
+        break;
+    default:
+        errMsg += "(UNKNOWN): ";
+        break;
+    }
+
+    switch(severity)
+    {
+        case GL_DEBUG_SEVERITY_HIGH:
+            level = LogLevel::FATAL;
+            break;
+        case GL_DEBUG_SEVERITY_MEDIUM:
+            level = LogLevel::ERROR;
+            break;
+        case GL_DEBUG_SEVERITY_LOW:
+            level = LogLevel::WARN;
+            break;
+        case GL_DEBUG_SEVERITY_NOTIFICATION:
+            level = LogLevel::INFO;
+            break;
+        default:
+            level = LogLevel::INFO;
+            break;
+    }
+
+    errMsg += message;
+
+    if(severity == GL_DEBUG_SEVERITY_HIGH)
+    {    
+        die(errMsg.c_str());
+    }
+
+    Log(level, errMsg.c_str());
+}
+
 using namespace sh3_gl;
 
 context::context(sh3_window& hwnd)
 {
-    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 3);
+    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 4);
```

Comment:
@z33ky I've changed the version of OpenGL we're using to 4.3 in this one as well (it's required for debugging and much better for compatibility). I don't feel the need for this to be in a separate PR, but I'll change it if you'd like :)

Reply:
It'd be awesome if we could target OpenGL ES 2.0 as a "minimum", which is supported on the Raspberry Pi and OpenPandora and probably a couple of other mobile devices.
Wikipedia says it's more or less OpenGL 2.0 without the fixed-function pipeline. Using a lower version core profile should enable us to have less issues if we or someone else ever tries to port our renderer to OpenGL ES 2.0. We could check for the `KHR_debug` extension instead (see `SDL_GL_ExtensionSupported` and `SDL_GL_GetProcAddress`) and just not install the callback if it doesn't exist (and log that).

Reply:
>It'd be awesome if we could target OpenGL ES 2.0 as a "minimum", which is supported on the Raspberry Pi and OpenPandora and probably a couple of other mobile devices.

> Wikipedia says it's more or less OpenGL 2.0 without the fixed-function pipeline. Using a lower version core profile should enable us to have less issues if we or someone else ever tries to port our renderer to OpenGL ES 2.0. We could check for the KHR_debug extension instead (see SDL_GL_ExtensionSupported and SDL_GL_GetProcAddress) and just not install the callback if it doesn't exist (and log that).

Should I do this in another PR? I'll get on it ASAP.

Reply:
> Would this be best to have some kind of preprocessor #define switch?

Either that or even a runtime mechanism, like function pointers that either point to an OpenGL or an OpenGL ES implementation.
For now I'd just stick with desktop OpenGL until we get stuff drawn and we have an idea of how we want to design our renderer.

> Should I do this in another PR? I'll get on it ASAP.

I'd stick with OpenGL 3.x core profile for now and just use the extension mechanism for this feature. This way we also don't use inefficient, deprecated stuff from before the "modern OpenGL".
When we get to needing more stuff from OpenGL 4.x we can reevaluate whether we want to keep using extensions or switch to 4.x proper.

> From the looks of it, trying to create an OpenGL 2.0 context on my PC (GTX470 Arch Linux) just defaults back to an OpenGL 4.6 context.

I just read that OpenGL vendors may return a compatible profile. Since you can use anything of OpenGL 2.0 with a 4.6 compatibility profile this is legitimate. Sadly this means we can unintentionally use features beyond what our specified version provides. Sigh.